dnsmasq: use the user defined dhcp-script directly when available

VincentRiou opened the pull request:
The introduction of hotplug scripts handling means that user defined scripts are not called directly when defined in config but sourced from dhcp-script.sh. This causes an issue when the dhcp script is not source-able (not an ash script). I added the user defined script in second position so that if both hotplug scripts and a user defined script exist, the current behaviour is maintained (I suspect there was a reason to source the file in the current version).

Signed-off-by: Vincent Riou <vincent@invizbox.com>

VincentRiou commented:
not sure if this needs someone tagged to be moved forward. (There's also a similar request against 18.06 - #1544)
I chose you @dedeckeh, as you seem to have reviewed similar patches, don't hesitate to let me know if this isn't the procedure.
Thanks!

jow- commented:
This fix looks wrong to me (or - depending on the opinion - at least incomplete). If a handler happens to be present (e.g. someone installs `/etc/hotplug.d/dhcp/99-whatever`) then the system dhcp script will be used again which then wrongly sources the user script again.

I think we should fix the invocation of the user script from the shipped DHCP script instead.

VincentRiou committed:
dnsmasq: use the user defined dhcp-script directly when available

The introduction of hotplug scripts handling means that user defined scripts are not called directly when defined in config but sourced from dhcp-script.sh. This causes an issue when the dhcp script is not source-able (not an ash script). I added the user defined script in second position so that if both hotplug scripts and a user defined script exist, the current behaviour is maintained (I suspect there was a reason to source the file in the current version).

Signed-off-by: Vincent Riou <vincent@invizbox.com>

VincentRiou commented:
@jow, my understanding is that what you're describing is the current expected behaviour i.e. someone who puts a handler **has** to put a source-able user-script in their dnsmasq instance. 
I'm not changing that expectation, I'm just bringing back the ability to use a user-script directly when there is no handler being set.
(added a handler validation on the addition of the environment variable)

Are you saying that the shipped DHCP script should try to source and if fail run the user script?

jow- commented:
No, sourcing the user script is an implementation error. While implementing the DHCP hotplug we didn't take the possibility into account that the user script is not a shell script but an arbitrary executable.

VincentRiou commented:
I see, thanks for the info. I'll have to have another look at it then as I was trying not to modify existing behaviour whilst bringing back the ability to run an arbitrary executable.

ynezz commented:
@VincentRiou Are you going to proceed with this PR or it could be closed?
